### PR TITLE
feat: refactor ws impl and add test framework

### PR DESCRIPTION
**Pull request by jensneuse**

### CODE REVIEW

File: router/core/websocket.go
Comment (StarpTech) [inline router/core/websocket.go:169]: Can we reuse it across requests?

File: router/core/websocket.go
Comment (StarpTech) [inline router/core/websocket.go:253]: I'd make it a pure function. No data is shared. Maybe we can even reuse it somewhere else.

File: router/core/flushwriter.go
Comment (StarpTech) [inline router/core/flushwriter.go:124]: I'd a comment. This is important to respond early to the upgrade right?

File: router/config/config.go
Comment (StarpTech) [inline router/config/config.go:153]: You made them configurable in yaml. Please update the docs.

File: router/config/config.go
Comment (StarpTech) [inline router/config/config.go:166]: Here too

File: router/config.yaml
Comment (StarpTech) [inline router/config.yaml:4]: Please keep it empty.

File: router/__schemas/compose.sh
Comment (StarpTech) [inline router/__schemas/compose.sh:1]: Those files are not required for the router. Can we move them to router_tests?

File: demo/cmd/subscriptiontest/run.sh
Comment (StarpTech) [inline demo/cmd/subscriptiontest/run.sh:6]: Could you document that a bit in the readme?

File: router-tests/events_test.go
Comment (StarpTech) [inline router-tests/events_test.go:86]: Lets use t.skip here.

File: router-tests/authentication_test.go
Comment (StarpTech) [inline router-tests/authentication_test.go:1]: Where you able to fix the races for websockets? Your PR does not test with `-race` enabled in the ci. You have to modify the Makefile in router_tests. There is a comment by me.

File: router-tests/events_test.go
Comment (StarpTech) [inline router-tests/events_test.go:354]: I'd always use `require.noError` because `assert` is handled optionally.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router-tests/module/module_test.go
Comment (StarpTech) [inline router-tests/module/module_test.go:1]: Why did you remove the test?

File: router-tests/module-jwt/module_test.go
Comment (StarpTech) [inline router-tests/module-jwt/module_test.go:1]: Why did you remove the test?

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router/core/flushwriter.go
Comment (jensneuse) [reply router/core/flushwriter.go:124]: delete

File: router/core/flushwriter.go
Comment (jensneuse) [reply router/core/flushwriter.go:126]: yes

File: router/core/websocket.go
Comment (jensneuse) [reply router/core/websocket.go:169]: yes, I think so but please test it

File: router/core/websocket.go
Comment (jensneuse) [reply router/core/websocket.go:253]: yes

File: router/__schemas/compose.sh
Comment (StarpTech) [reply router/__schemas/compose.sh:1]: I kept them and document it.

File: router-tests/authentication_test.go
Comment (StarpTech) [reply router-tests/authentication_test.go:1]: Works. Tested locally.

File: router/core/websocket.go
Comment (StarpTech) [reply router/core/websocket.go:169]: I skipped this optimization because there is no easy way to access the protocol afterwards.

**Comment by devsergiy** [comment]
@StarpTech please do not merge this when it's done

I will need to do a hotfix on the engine before merging this

**Review by StarpTech — APPROVED** [review]
LGTM

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**
